domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java

```diff
@@ -1596,91 +1595,207 @@ public TotaledResults(ResultSet set, Integer total) {
+    if (connResult.failed()) {
+      log.error(connResult.cause().getMessage(), connResult.cause());
+      replyHandler.handle(Future.failedFuture(connResult.cause()));
+      return;
+    }
+    this.doStreamGetCount(connResult.result(), table, clazz, fieldName, wrapper, returnIdField,
+      distinctOn, facets, replyHandler);
+  }
 
-            Map<String, Method> externalColumnSetters = getExternalColumnSetters(resultsHelper.columnNames,
-                resultsHelper.clazz, isAuditFlavored);
+  /**
+   * private for now, might be public later (and renamed)
+   * @param <T>
+   * @param connection
+   * @param table
+   * @param clazz
+   * @param fieldName
+   * @param wrapper
+   * @param returnIdField
+   * @param distinctOn
+   * @param facets
+   * @param replyHandler
+   */
+  @SuppressWarnings({"unchecked", "squid:S00107"})    // Method has >7 parameters
+  private <T> void doStreamGetCount(SQLConnection connection,
+    String table, Class<T> clazz, String fieldName, CQLWrapper wrapper,
+    boolean returnIdField, String distinctOn, List<FacetField> facets,
+    Handler<AsyncResult<PostgresClientStreamResult>> replyHandler) {
 
-            sqlRowStream.resultSetClosedHandler(v -> sqlRowStream.moreResults()).handler(r -> {
-              JsonObject row = convertRowStreamArrayToObject(sqlRowStream, r);
-              try {
-                streamHandler.handle((T) deserializeRow(resultsHelper, externalColumnSetters, isAuditFlavored, row));
-              } catch (Exception e) {
-                sqlRowStream.close();
-                if (!transactionMode) {
-                  connection.close();
-                }
-                log.error(e.getMessage(), e);
-                replyHandler.handle(Future.failedFuture(e));
-              }
-            }).endHandler(v2 -> {
-              if (!transactionMode) {
-                connection.close();
-              }
-              replyHandler.handle(Future.succeededFuture());
-            });
+    try {
+      QueryHelper queryHelper = buildQueryHelper(table,
+        fieldName, wrapper, returnIdField, facets, distinctOn);
+
+      connection.querySingle(queryHelper.countQuery, countQueryResult -> {
+        if (countQueryResult.failed()) {
+          replyHandler.handle(Future.failedFuture(countQueryResult.cause()));
+          return;
+        }
+        ResultInfo resultInfo = new ResultInfo();
+        resultInfo.setTotalRecords(countQueryResult.result().getInteger(0));
+        doStreamGetQuery(connection, queryHelper.selectQuery, resultInfo,
+          clazz, facets, replyHandler);
+      });
+    } catch (Exception e) {
+      log.error(e.getMessage(), e);
+      replyHandler.handle(Future.failedFuture(e));
+    }
+  }
 
+  private <T> void doStreamGetQuery(SQLConnection connection, String selectQuery,
+    ResultInfo resultInfo, Class<T> clazz, List<FacetField> facets,
+    Handler<AsyncResult<PostgresClientStreamResult>> replyHandler) {
```

Comment:
Can you replace `<PostgresClientStreamResult>` by `<PostgresClientStreamResult<T>>`?